Quadlet - new unit type .login (#26507)
containers/podman: docs/source/markdown/podman-systemd.unit.5.md

Path of the authentication file.

> This is equivalent to the `--authfile` option.

Reviewer: ```suggestion
This is equivalent to the Podman `--authfile` option.
```
I'm not sure why "Podman" was dropped here, but not in other places?

| LogoutOnStop=true                               | Add ExecStopPost to log out of the registry when the unit is stopped |
| Password=mypassword                             | --password-stdin and set StandardInputText                           |
| PodmanArgs=--compat-auth-file=/etc/compat\.json | --compat-auth-file=/etc/registry/compat\.json                        |
| Registry=quai.io                                | podman login quai.io                                                 |
| Secret=mysecret                                 | --secret=mysecret                                                    |
| ServiceName=name                                | Name the systemd unit `name.service`                                 |
| TLSVerify=false                                 | --tls-verify=false                                                   |
| Username=myuser                                 | --username=myuser                                                    |

### `AuthFile=`

Path of the authentication file.

This field is mandatory when linking between an `AuthFile` key of a different unit and the `.login` unit.

> This is equivalent to the `--authfile` option.

Reviewer: ```suggestion
This is equivalent to the Podman `--authfile` option.
```

In some places, you have `Podman` for a line like this, in others, you do not.  I prefer having Podman for each.  However, your call, I'd just ask for consistency.

### `CertDir=`

Use certificates at path (*.crt, *.cert, *.key) to connect to the registry.

This is equivalent to the Podman `--cert-dir` option.

### `LogoutOnStop=` (defaults to `false`)

When set to `true` the logout will be called when the service is stopped

### `Password=`

Use the password to connect to the registry.
In order to prevent password leak, instead of using the `--password` option,
Quadlet will set `--password-stdin` and use systemd `StandardInputText` to pass the data.

### `PodmanArgs=`

This key contains a list of arguments passed directly to the end of the `podman login` command
in the generated file. It can be used to access Podman features otherwise unsupported by the generator. Since the generator is unaware
of what unexpected interactions can be caused by these arguments, is not recommended to use
this option.

The format of this is a space separated list of arguments, which can optionally be individually

> escaped to allow inclusion of whitespace and other control characters.

Reviewer: It would be nice to add some examples at the bottom of this page.  Not a necessity.

This key can be listed multiple times.

### `Registry=`

Use the URL of the registry to connect to. This key is mandatory.

This is equivalent to the last argument of `podman login`.

### `Secret=`

> Use a podman secret for the credentials

Reviewer: ```suggestion
Use a podman secret for the credentials.
```

This is equivalent to the Podman `--variant` option.

## Login units [Login]

Login files are named with a `.login` extension and contain a section `[Login]` describing the
login command. The generated service is a one-time command that logs into a registry.

> Using login units allows pulling images from private registries without having to login manually.

Reviewer 2: A note about Systemd unit ordering should be added here (i.e. that users must ensure that the `.login` unit runs before any `.container` unit that wants to use these credentials starts - either manually, or by using `AuthFile=private-repo.login` in the `.container` unit)

Reviewer 2: Even though it might be a little off-topic, hinting users into the right direction might be helpful. How about something like the following:

> Note that Systemd generally starts units in parallel. This can cause ordering issues, because a `.container` unit can't pull images from an authenticated source unless the `.login` unit with the necessary credentials runs already. Quadlet automatically ensures proper order when adding `AuthFile=private-repo.login` to a `.container` unit. However, else one must either add proper `After=` directives (e.g. `After=private-repo-login.service`) to the `.container` units manually, or use a custom synchronisation point like creating a `containers.target` unit and adding `Before=containers.target` to all `.login` units and `After=containers.target` to all `.container` units.

There is only one required key, `Registry`, which defines the URL of the registry to log into.

To avoid password leaking, Quadlet will set the permissions of the generated service file as 0600.
It is recommended to do the same for `.login` unit files.

Valid options for `[Login]` are listed below:

| **[Login] options**                             | **podman login**                                                     |
|-------------------------------------------------|----------------------------------------------------------------------|
| AuthFile=/etc/registry/auth\.json               | --authfile=/etc/registry/auth\.json                                  |
| CertDir=/etc/certs                              | --cert-dir=/etc/certs                                                |
| LogoutOnStop=true                               | Add ExecStopPost to log out of the registry when the unit is stopped |

> | Password=mypassword                             | --password-stdin and set StandardInputText                           |

Reviewer 3: Can we just… not do that?

Author: I don't think Quadlet should block users from doing what podman allows them to

| Registry=quai.io                                | podman login quai.io                                                 |
| Secret=mysecret                                 | --secret=mysecret                                                    |
| ServiceName=name                                | Name the systemd unit `name.service`                                 |
| TLSVerify=false                                 | --tls-verify=false                                                   |
| Username=myuser                                 | --username=myuser                                                    |

### `AuthFile=`

Path of the authentication file.

This field is mandatory when linking between an `AuthFile` key of a different unit and the `.login` unit.
Do not use unit specific systemd specifiers (e.g. `%n`) as they will be resolved differently in the dependent units.

This is equivalent to the Podman `--authfile` option.

### `CertDir=`

Use certificates at path (*.crt, *.cert, *.key) to connect to the registry.

This is equivalent to the Podman `--cert-dir` option.

### `LogoutOnStop=` (defaults to `false`)

> When set to `true`, the logout will be called when the service is stopped.

Reviewer 2: ```suggestion
When set to `true`, the `podman logout` will be called when the service stops.
```

### `Password=`

Use the password to connect to the registry.
In order to prevent password leak, instead of using the `--password` option,
Quadlet will set `--password-stdin` and use systemd `StandardInputText` to pass the data.

### `PodmanArgs=`

This key contains a list of arguments passed directly to the end of the `podman login` command
in the generated file. It can be used to access Podman features otherwise unsupported by the generator.
Since the generator is unaware of the unexpected interactions that can be caused by these arguments,
it is not recommended to use this option.

The format of this is a space separated list of arguments, which can optionally be individually
escaped to allow inclusion of whitespace and other control characters.

This key can be listed multiple times.

### `Registry=`

Use the URL of the registry to connect to. This key is mandatory.

This is equivalent to the last argument of `podman login`.

### `Secret=`

> Use a podman secret for the credentials.

Reviewer: ```suggestion
Use a Podman secret for the credentials.
```

| CertDir=/etc/certs                              | --cert-dir=/etc/certs                                                |
| LogoutOnStop=true                               | Add ExecStopPost to log out of the registry when the unit is stopped |
| Password=mypassword                             | --password=mypassword                                                |
| PodmanArgs=--compat-auth-file=/etc/compat\.json | --compat-auth-file=/etc/registry/compat\.json                        |
| Registry=quai.io                                | podman login quai.io                                                 |
| Secret=mysecret                                 | --secret=mysecret                                                    |
| ServiceName=name                                | Name the systemd unit `name.service`                                 |
| TLSVerify=false                                 | --tls-verify=false                                                   |
| Username=myuser                                 | --username=myuser                                                    |

### `AuthFile=`

Path of the authentication file.

> This field is mandatory when linking between an `AuthFile` key of a different unit and the `.login` unit.

Author: Good question, I'm not sure. 
Not setting one, means that the login information is stored in the default file. So, units can depend on on the `.login` unit (in the `Wants` and `After`) without setting `AuthKey`.
But, maybe it is cleaner

> This field is mandatory when linking between an `AuthFile` key of a different unit and the `.login` unit.

Author: My thought was that if the `.login` file doesn't specify an `AuthFile` then login info will be stored in the default file. So, the `.container` does not need to specify an `AuthFile` either, but it will need to specify the dependecy.
As for `%n`, you are correct. I'll add it to the doc

> This field is mandatory when linking between an `AuthFile` key of a different unit and the `.login` unit.

Reviewer 4: > Auth files don't have to be unique and can store the credentials of multiple registries (i.e. using the default auth file is totally file).

I think the main issue here is AFAICS there is no locking for writing the authfile. So if you actually use the same file in many login units they would all race against each other and potentially overwrite a previous login causing hard to find race conditions.

> This field is mandatory when linking between an `AuthFile` key of a different unit and the `.login` unit.

Reviewer 2: This got me thinking: I don't know Quadlet's code, but since `AuthFile=private-repo.login` in `.container` units work, Quadlet must be aware of other Quadlet units when generating the actual Systemd units. So, how about grouping `.login` units that attempt to write to the same auth file together and adding `After=` directives to the generated Systemd units to ensure that they run sequentially? This would actually solve the issue with race conditions.
